src/jdk.incubator.code/share/classes/jdk/incubator/code/type/CoreTypeFactory.java

```diff
@@ -90,168 +93,11 @@ public TypeElement constructType(TypeElement.ExternalizedTypeElement tree) {
-                // the inner class name is obtained by subtracting the name of the enclosing type
-                // from the name of the inner type (and also dropping an extra '$')
-                String innerName = innerType.toNominalDescriptor().displayName()
-                        .substring(enclType.toNominalDescriptor().displayName().length() + 1);
-                JavaType qual = JavaType.qualified(enclType, innerName);
-                return (innerType.hasTypeArguments()) ?
-                    JavaType.parameterized(qual, innerType.typeArguments()) : qual;
-            } else {
-                // primitive or reference
-                JavaType t = switch (identifier) {
-                    case "boolean" -> JavaType.BOOLEAN;
-                    case "byte" -> JavaType.BYTE;
-                    case "char" -> JavaType.CHAR;
-                    case "short" -> JavaType.SHORT;
-                    case "int" -> JavaType.INT;
-                    case "long" -> JavaType.LONG;
-                    case "float" -> JavaType.FLOAT;
-                    case "double" -> JavaType.DOUBLE;
-                    case "void" -> JavaType.VOID;
-                    default -> JavaType.type(ClassDesc.of(identifier));
-                };
-                if (!tree.arguments().isEmpty()) {
-                    if (t instanceof PrimitiveType) {
-                        throw new IllegalArgumentException("primitive type: " + tree);
-                    }
-                    return JavaType.parameterized(t,
-                            tree.arguments().stream().map(ete -> (JavaType) constructType(ete)).toList());
-                } else {
-                    return t;
-                }
-            }
-        }
-
-        static IllegalArgumentException badType(ExternalizedTypeElement tree, String str) {
-            return new IllegalArgumentException(String.format("Bad %s: %s", str, tree));
-        }
-
-        private JavaType constructTypeArgument(ExternalizedTypeElement element, int index, Predicate<JavaType> filter) {
-            ExternalizedTypeElement arg = element.arguments().get(index);
-            JavaType type = (JavaType) constructType(arg);
-            if (!filter.test(type)) {
-                throw new IllegalArgumentException(String.format("Unexpected argument %s", element));
-            } else {
-                return type;
-            }
-        }
-
-        private static Predicate<JavaType> NO_WILDCARDS = t -> !(t instanceof WildcardType);
-        private static Predicate<JavaType> CLASS = t -> t instanceof ClassType;
-    };
+    public static final TypeElementFactory JAVA_TYPE_FACTORY = tree -> switch (JavaTypeUtils.Kind.of(tree)) {
+            case INFLATED_TYPE -> JavaTypeUtils.toJavaType(tree);
+            case INFLATED_REF -> JavaTypeUtils.toJavaRef(tree);
+            default -> throw new UnsupportedOperationException("Unsupported: " + tree);
```

Comment:
I'm starting to think this should return `null`, as we used that as "composition" signal?